src/ae.c

```diff
@@ -272,6 +278,60 @@ static int64_t usUntilEarliestTimer(aeEventLoop *eventLoop) {
     return (now >= earliest->when) ? 0 : earliest->when - now;
 }
 
+static int processFileEvents(aeEventLoop *eventLoop, int numevents) {
+    for (int j = 0; j < numevents; j++) {
+        int fd = eventLoop->fired[j].fd;
+        aeFileEvent *fe = &eventLoop->events[fd];
+        int mask = eventLoop->fired[j].mask;
+        int fired = 0; /* Number of events fired for current fd. */
+
+        /* Normally we execute the readable event first, and the writable
+         * event later. This is useful as sometimes we may be able
+         * to serve the reply of a query immediately after processing the
+         * query.
+         *
+         * However if AE_BARRIER is set in the mask, our application is
+         * asking us to do the reverse: never fire the writable event
+         * after the readable. In such a case, we invert the calls.
+         * This is useful when, for instance, we want to do things
+         * in the beforeSleep() hook, like fsyncing a file to disk,
+         * before replying to a client. */
+        int invert = fe->mask & AE_BARRIER;
+
+        /* Note the "fe->mask & mask & ..." code: maybe an already
+         * processed event removed an element that fired and we still
+         * didn't processed, so we check if the event is still valid.
+         *
+         * Fire the readable event if the call sequence is not
+         * inverted. */
+        if (!invert && fe->mask & mask & AE_READABLE) {
+            fe->rfileProc(eventLoop, fd, fe->clientData, mask);
+            fired++;
+            fe = &eventLoop->events[fd]; /* Refresh in case of resize. */
+        }
+
+        /* Fire the writable event. */
+        if (fe->mask & mask & AE_WRITABLE) {
+            if (!fired || fe->wfileProc != fe->rfileProc) {
+                fe->wfileProc(eventLoop, fd, fe->clientData, mask);
+                fired++;
+            }
+        }
+
+        /* If we have to invert the call, fire the readable event now
+         * after the writable one. */
+        if (invert) {
+            fe = &eventLoop->events[fd]; /* Refresh in case of resize. */
+            if ((fe->mask & mask & AE_READABLE) &&
+                (!fired || fe->wfileProc != fe->rfileProc)) {
+                fe->rfileProc(eventLoop, fd, fe->clientData, mask);
+            }
+        }
+    }
+
+    return numevents;
+}
+
```

Comment:
This function is extracted from ```aeProcessEvents()```.

1. The ```fired++``` at the end of the loop (line 462 of the original code) be removed, because the ```fired``` will always be reinitialized on the next loop.
2. The```processed``` be removed. Once ```processFileEvent()``` be called, currently it seems that ```processed``` is always equal to ```numevens```, so there is no need to accumulate any more.

This function was extracted because I think it improves the readability of the code.
1. The original ```aeProcessEvent``` function is a bit long
2. ```processFileEvents``` is actually at the same level of abstraction as ```processTimeEvents```

So, in my opinion. After extracting ```processFileEvents```, the logic of ```aeProcessEvent``` is clearer.
Maybe, I should move this refactoring to another commit?

Reply:
i'd like to re-introduce that last `filed++`, despite doing nothing, it think it serves to make a statement.